flang/lib/Optimizer/Transforms/ArrayValueCopy.cpp

```diff
@@ -101,80 +101,136 @@ class ArrayCopyAnalysis {
       popFn(rop);
       return;
     }
-    if (auto rop = mlir::dyn_cast<fir::IfOp>(op)) {
-      popFn(rop);
-      return;
+    // Otherwise, a block argument is provided via the pred blocks.
+    for (auto *pred : ba.getOwner()->getPredecessors()) {
+      auto u = pred->getTerminator()->getOperand(ba.getArgNumber());
+      collectArrayAccessFrom(u);
     }
+  }
 
-    // Otherwise, Op does not contain a region so just chase its operands.
-    if (mlir::isa<ArrayLoadOp, ArrayUpdateOp, ArrayFetchOp>(op)) {
-      LLVM_DEBUG(llvm::dbgs() << "add " << *op << " to reachable set\n");
-      reach.emplace_back(op);
+private:
+  /// Is \op inside the loop nest region ?
+  bool opIsInsideLoops(mlir::Operation *op) const {
+    auto *region = op->getParentRegion();
+    while (region) {
+      if (region == loopRegion)
+        return true;
+      region = region->getParentRegion();
     }
-    for (auto u : op->getOperands())
-      populateSets(reach, visited, u);
-    return;
+    return false;
   }
 
-  // Process a block argument.
-  auto ba = val.cast<mlir::BlockArgument>();
-  auto *parent = ba.getOwner()->getParentOp();
-  // If inside an Op holding a region, the block argument corresponds to an
-  // argument passed to the containing Op.
-  auto popFn = [&](auto rop) {
-    populateSets(reach, visited, rop.blockArgToSourceOp(ba.getArgNumber()));
-  };
-  if (auto rop = mlir::dyn_cast<DoLoopOp>(parent)) {
-    popFn(rop);
-    return;
-  }
-  if (auto rop = mlir::dyn_cast<IterWhileOp>(parent)) {
-    popFn(rop);
-    return;
-  }
-  // Otherwise, a block argument is provided via the pred blocks.
-  for (auto *pred : ba.getOwner()->getPredecessors()) {
-    auto u = pred->getTerminator()->getOperand(ba.getArgNumber());
-    populateSets(reach, visited, u);
+  void collectValuesStoredTo(mlir::Operation *alloc) {}
+
+  /// Recursively trace the use of an operation results, calling
+  /// collectArrayAccessFrom on the direct and indirect user operands.
+  void followUsers(mlir::Operation *op) {
+    for (auto userOperand : op->getOperands())
+      collectArrayAccessFrom(userOperand);
+    // Go through potential converts/coordinate_op.
+    for (auto indirectUser : op->getUsers())
+      followUsers(indirectUser);
   }
-}
+
+  llvm::SmallVectorImpl<mlir::Operation *> &reach;
+  llvm::SmallPtrSet<mlir::Value, 16> visited;
+  /// Region of the loops nest that produced the array value.
+  mlir::Region *loopRegion;
+};
+} // namespace
 
 /// Return all ops that produce the array value that is stored into the
 /// `array_merge_store`.
 static void reachingValues(llvm::SmallVectorImpl<mlir::Operation *> &reach,
                            mlir::Value seq) {
   reach.clear();
-  llvm::SmallPtrSet<mlir::Value, 16> visited;
-  populateSets(reach, visited, seq);
+  mlir::Region *loopRegion = nullptr;
+  if (auto doLoop = mlir::dyn_cast_or_null<fir::DoLoopOp>(seq.getDefiningOp()))
+    loopRegion = &doLoop->getRegion(0);
```

Comment:
Not sure about this new restriction (with the do_loop). We can discuss offline though.